flox/core.py

```diff
@@ -1550,25 +1616,39 @@ def _factorize_multiple(by, expected_groups, any_by_dask, reindex):
             *by_,
             chunks=tuple(chunks.values()),
             meta=np.array((), dtype=np.int64),
-            **kwargs,
-        )
-        found_groups = tuple(
-            None if is_duck_dask_array(b) else pd.unique(b.reshape(-1)) for b in by
-        )
-        grp_shape = tuple(
-            len(e) if e is not None else len(f) for e, f in zip(expected_groups, found_groups)
+            axes=(),  # always (), we offset later if necessary.
+            expected_groups=expected_groups,
+            fastpath=True,
+            reindex=reindex,
         )
-    else:
-        group_idx, found_groups, grp_shape = factorize_(by, **kwargs)
 
-    final_groups = tuple(
-        found if expect is None else expect.to_numpy()
-        for found, expect in zip(found_groups, expected_groups)
-    )
+        fg, gs = [], []
+        for b, e in zip(by, expected_groups):
+            if e is None:
+                if is_duck_dask_array(b):
+                    raise ValueError(
+                        "Please provide expected_groups when grouping by a dask array."
+                    )
+
+                f = pd.unique(b.reshape(-1))
+            else:
+                f = e.to_numpy()
+
+            fg.append(f)
+            gs.append(len(f))
+
+        found_groups = tuple(fg)
+        grp_shape = tuple(gs)
```

Comment:
I prefer the previous comprehension.

Reply:
Previous comprehension didn't have the same typing for `found_groups` as `factorize_` and `grp_shape` could crash (`len(None)`).